// tweeter-shared/src/index.ts
export {Follow} from "./model/domain/Follow";
export {PostSegment, Type} from "./model/domain/PostSegment";
export {Status} from "./model/domain/Status";
export {User} from "./model/domain/User";
export {AuthToken} from "./model/domain/AuthToken";

// All classes that should be avaialble to other modules need to exported here. export * does not work when 
// uploading to lambda. Instead we have to list each export.
export {FakeData} from "./util/FakeData";
export {
    TweeterRequest,
    LoginRequest,
    RegisterRequest,
    LogoutRequest,
    GetUserRequest,
    LoadMoreStatusItemsRequest,
    PostStatusRequest,
    FollowerOperationsRequest,
    GetIsFollowerRequest,
    LoadMoreUserItemsRequest
} from "./model/net/Request";
export {
    TweeterResponse,
    AuthenticateResponse,
    GetUserResponse,
    VoidResponse,
    LoadMoreStatusItemsResponse,
    GetFollowCountResponse,
    GetIsFollowerResponse,
    LoadMoreUserItemsResponse,
    FollowActionResponse
} from "./model/net/Response";
